cpp/src/arrow/builder.cc

```diff
@@ -1110,20 +1034,202 @@ Status DictionaryBuilder<T>::AppendDictionary(const Scalar& value) {
+template class UniqueBuilder<TimestampType>;
+template class UniqueBuilder<FloatType>;
+template class UniqueBuilder<DoubleType>;
+template class UniqueBuilder<FixedSizeBinaryType>;
+template class UniqueBuilder<BinaryType>;
+template class UniqueBuilder<StringType>;
+
+// ----------------------------------------------------------------------
+// DictionaryBuilder
+
+template <typename T>
+DictionaryBuilder<T>::DictionaryBuilder(const std::shared_ptr<DataType>& type,
+                                        MemoryPool* pool)
+    : ArrayBuilder(type, pool), unique_builder_(type, pool), values_builder_(pool) {}
+
+DictionaryBuilder<NullType>::DictionaryBuilder(const std::shared_ptr<DataType>& type,
+                                               MemoryPool* pool)
+    : ArrayBuilder(type, pool), values_builder_(pool) {}
+
+DictionaryBuilder<NullType>::~DictionaryBuilder() {}
+
+template <>
+DictionaryBuilder<FixedSizeBinaryType>::DictionaryBuilder(
+    const std::shared_ptr<DataType>& type, MemoryPool* pool)
+    : ArrayBuilder(type, pool), unique_builder_(type, pool), values_builder_(pool) {}
+
+template <typename T>
+Status DictionaryBuilder<T>::Init(int64_t elements) {
+  RETURN_NOT_OK(ArrayBuilder::Init(elements));
+  RETURN_NOT_OK(unique_builder_.Init(elements));
+  return values_builder_.Init(elements);
+}
+
+Status DictionaryBuilder<NullType>::Init(int64_t elements) {
+  RETURN_NOT_OK(ArrayBuilder::Init(elements));
+  return values_builder_.Init(elements);
+}
+
+template <typename T>
+Status DictionaryBuilder<T>::Resize(int64_t capacity) {
+  if (capacity < kMinBuilderCapacity) {
+    capacity = kMinBuilderCapacity;
+  }
+
+  if (capacity_ == 0) {
+    return Init(capacity);
+  } else {
+    RETURN_NOT_OK(unique_builder_.Resize(capacity));
+    return ArrayBuilder::Resize(capacity);
+  }
+}
+
+Status DictionaryBuilder<NullType>::Resize(int64_t capacity) {
+  if (capacity < kMinBuilderCapacity) {
+    capacity = kMinBuilderCapacity;
+  }
+
+  if (capacity_ == 0) {
+    return Init(capacity);
+  } else {
+    return ArrayBuilder::Resize(capacity);
+  }
+}
+
+template <typename T>
+Status DictionaryBuilder<T>::FinishInternal(std::shared_ptr<ArrayData>* out) {
+  std::shared_ptr<Array> dictionary;
+  RETURN_NOT_OK(unique_builder_.Finish(&dictionary));
+
+  RETURN_NOT_OK(values_builder_.FinishInternal(out));
+  (*out)->type = std::make_shared<DictionaryType>((*out)->type, dictionary);
+  return Status::OK();
+}
+
+Status DictionaryBuilder<NullType>::FinishInternal(std::shared_ptr<ArrayData>* out) {
+  std::shared_ptr<Array> dictionary = std::make_shared<NullArray>(0);
+
+  RETURN_NOT_OK(values_builder_.FinishInternal(out));
+  (*out)->type = std::make_shared<DictionaryType>((*out)->type, dictionary);
+  return Status::OK();
+}
+
+template <typename T>
+Status DictionaryBuilder<T>::Append(const Scalar& value) {
+  RETURN_NOT_OK(Reserve(1));
+  int32_t index;
+  RETURN_NOT_OK(unique_builder_.Append(value, &index));
```

Comment:
In my experience with implementing these things in the past, we are probably going to want to create a base hash table pass class that accepts a functor whose action can be inlined in the innermost loop of the hash table pass. This functor would accumulate the results of the hash table pass, whether it's one of:

* isin (values_to_test, value_set) -> boolean array
* match (values_to_test, value_set) -> integer array (positional version of isin)
* unique (values) -> unique values
* dictionary-encode / factorize (values) -> unique values + category indices
* value-counts(values) -> unique values + counts

We should write functors for each of these "actions" and then instantiate kernels for each input type and functor.

On this note, probably all of this code (including the DictionaryBuilder code -- such that it does not disrupt parquet-cpp) should move to src/arrow/compute.

I also question whether we need to preserve a scalar append API when instead we could do hash table passes in batches in the places where incremental dictionary-encoding is important

Reply:
As part of this we need to also define a stateful API so that we can keep pushing more data through the kernel for the ones that need to maintain state. Some of the "modes" of the hash table will write their results into output memory, where others (e.g. unique) will keep accumulating the result